controllers/purge_controller.go

```diff
@@ -19,180 +19,195 @@ package controllers
 
-type CRDMatcher func(crd apiextensions.CustomResourceDefinition) bool
-
-// PurgeReconciler reconciles a Kyma object.
 type PurgeReconciler struct {
 	client.Client
 	record.EventRecorder
 	ResolveRemoteClient   RemoteClientResolver
 	PurgeFinalizerTimeout time.Duration
-	SkipCRDs              CRDMatcher
+	SkipCRDs              matcher.CRDMatcherFunc
 	IsManagedKyma         bool
 }
 
+//nolint:funlen
 func (r *PurgeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	logger := ctrlLog.FromContext(ctx)
 	logger.V(log.InfoLevel).Info("Purge Reconciliation started")
 
 	ctx = adapter.ContextWithRecorder(ctx, r.EventRecorder)
 
-	// check if kyma resource exists
 	kyma := &v1beta2.Kyma{}
 	if err := r.Get(ctx, req.NamespacedName, kyma); err != nil {
-		if !util.IsNotFound(err) {
+		if util.IsNotFound(err) {
 			logger.V(log.DebugLevel).Info(fmt.Sprintf("Kyma %s not found, probably already deleted", req.NamespacedName))
 			return ctrl.Result{}, fmt.Errorf("purgeController: %w", err)
 		}
 		return ctrl.Result{Requeue: false}, nil
 	}
 
 	if kyma.DeletionTimestamp.IsZero() {
-		if err := r.EnsurePurgeFinalizer(ctx, kyma); err != nil {
+		if err := r.ensurePurgeFinalizer(ctx, kyma); err != nil {
 			logger.V(log.DebugLevel).Info(fmt.Sprintf("Updating purge finalizers for Kyma  %s failed with err %s",
 				req.NamespacedName, err))
 			return ctrl.Result{}, err
 		}
 		return ctrl.Result{Requeue: true}, nil
 	}
 
-	// condition to check if deletionTimestamp is set, retry until it gets fully deleted
 	deletionDeadline := kyma.DeletionTimestamp.Add(r.PurgeFinalizerTimeout)
+	if time.Now().Before(deletionDeadline) {
+		requeueAfter := time.Until(deletionDeadline.Add(time.Second))
+		logger.V(log.DebugLevel).Info(fmt.Sprintf("Purge reconciliation for Kyma  %s will be requeued after %s",
+			req.NamespacedName, requeueAfter))
+		return ctrl.Result{RequeueAfter: requeueAfter}, nil
+	}
 
-	if time.Now().After(deletionDeadline) { //nolint:nestif
-		remoteClient, err := r.ResolveRemoteClient(ctx, client.ObjectKeyFromObject(kyma))
-		if util.IsNotFound(err) {
-			if err := r.DropPurgeFinalizer(ctx, kyma); err != nil {
-				logger.Error(err, "Couldn't remove Purge Finalizer from the Kyma object")
-				return ctrl.Result{}, err
+	start := time.Now()
+	remoteClient, err := r.ResolveRemoteClient(ctx, client.ObjectKeyFromObject(kyma))
+	if util.IsNotFound(err) {
+		if err := r.dropPurgeFinalizer(ctx, kyma); err != nil {
+			logger.Error(err, "Couldn't remove Purge Finalizer from the Kyma object")
+			if err := metrics.UpdatePurgeError(kyma, metrics.ErrPurgeFinalizerRemoval); err != nil {
+				logger.Error(err, "Failed to update error metrics")
 			}
-			return ctrl.Result{Requeue: true}, nil
-		}
-
-		if err != nil {
 			return ctrl.Result{}, err
 		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
```

Comment:
```suggestion
		return ctrl.Result{}, err
```
If has err, it will requeue anyway